### DISCUSSION TITLE
i think there might be a problem with WMT.requestMIDIAccess when testing react components [jsdom]

### DISCUSSION THREAD

**Discussion by glend1** [jazz-soft/web-midi-test#6] [Q&A]
my test is as follows

````
it("testing purposes only", async () => { 
            navigator.requestMIDIAccess = WMT.requestMIDIAccess
            const { result, waitForNextUpdate } = renderHook(() => {return MidiConnection()})
            act(() => {
                result.current.enable()
            })
            await waitForNextUpdate();
            expect(result.current.error).toBe("failed")
          })
````

`MidiConnection()` uses [WebMidi](https://github.com/djipco/webmidi/) the code for which is;

````
export function MidiConnection() {
    const {array: midiInputs, push: midiPush, filter: midiFilter} = useArray(["none"])
    const [error, setError] = useState<string | undefined>();
    function enable() {
        WebMidi.addListener("connected", (e) => { if (isInput(e)) {midiPush(e.port.name)}});
        WebMidi.addListener("disconnected", (e) => {
            e.port.removeListener()
            if (isInput(e)) {midiFilter((str) => {return str != e.port.name})}
        });
        WebMidi.
        enable().
        catch((err) => {
            setError(err.message)
        })
    }
    return ({ports: midiInputs, error, enable})
}
````

when i run this i get the following error, `Timed out in waitForNextUpdate after 1000ms.`

when i substitute out the `requestMIDIAccess` assignment to 

````
navigator.requestMIDIAccess = (options?: WebMidi.MIDIOptions | undefined) => {return pro};
````
this is `pro`

````
export const pro = new Promise((resolve, reject) => {
  if (1 == 1) {
    resolve("resolved");
  } else {
    reject("failed");
  }
});
````

my test fails, but on the assertion, `expect(received).toBe(expected) // Object.is equality`.
importantly, the value of `result.current.error` is not the default value in the debugger so the state is updating

I thought it might be the hook im trying to render. so i changed the hook

````
export function CHook() {
  const [error, setError] = useState();
  function enable() {
    pro
      .then((msg) => {
        setError(msg);
      })
      .catch((msg) => {
        setError(msg);
      });
  }
  return { error, setError, enable };
}
````

functionally this should be the same and mocks out midi. I get the same inequality assertion failure.
importantly, the value of `result.current.error` is not the default value in the debugger so the state is updating.

it appears the only thing im doing differently between all my tests which behave different is and its appear to just stall making the test time out. do you know why this might be happening?

**Comment by jazz-soft** [comment]
I'm not sure if it relates to your issue, but it looks like all your tests use the same object WMT.requestMIDIAccess and if its state is changed in one test, its behavior may change in another test.

**Comment by glend1** [comment]
no, its the only test that is running

**Comment by jazz-soft** [comment]
You use some products that I am not well familiar with...
Could you please make a minimal stand alone test for me to debug?

**Comment by glend1** [comment]
is a [codesandbox](https://codesandbox.io/s/cocky-dirac-db53z5?file=/src/App.js) good?

**Reply by jazz-soft** [reply]
Never used it before, but let's try...

**Reply by glend1** [reply]
there is an option to export as zip from the menu if it becomes cumbersome.
